napari/_qt/_tests/test_plugin_widgets.py

```diff
@@ -195,3 +133,65 @@ def test_inject_viewer_proxy(make_napari_viewer):
     with patch('napari.utils.misc.ROOT_DIR', new='/some/other/package'):
         with pytest.warns(FutureWarning):
             wdg.fail()
+
+
+@pytest.mark.parametrize(
+    "widget_callable, param",
+    [
+        (QWidget_example, 'napari_viewer'),
+        (QWidget_string_annnot, 'test'),
+        (Container_example, 'test'),
+    ],
+)
+def test_get_widget_viewer_param(widget_callable, param):
+    """Test `_get_widget_viewer_param` returns correct parameter name."""
+    out = _get_widget_viewer_param(widget_callable, 'widget_name')
+    assert out == param
+
+
+def test_get_widget_viewer_param_error():
+    """Test incorrect subclass raises error in `_get_widget_viewer_param`."""
+    with pytest.raises(TypeError) as e:
+        _get_widget_viewer_param(ErrorWidget, 'widget_name')
+    assert "'widget_name' must be `QtWidgets.QWidget`" in str(e)
+
+
+def test_widget_hide_destroy(make_napari_viewer):
+    """Test that widget hide and destroy works."""
+    viewer = make_napari_viewer()
+    viewer.window.add_dock_widget(QWidget_example(), name='test')
+    widget = viewer.window._dock_widgets['test']
+
+    # Check widget persists after hide
+    widget.title.hide_button.click()
+    assert widget
+    # Check that widget removed from `_dock_widgets` dict when closed
+    widget.destroyOnClose()
+    assert 'test' not in viewer.window._dock_widgets
+
+
+@pytest.mark.parametrize(
+    "Widget",
+    [
+        QWidget_example,
+        Container_example,
+        magic_widget_example,
+        callable_example,
+    ],
+)
+def test_widget_types_supported(
+    make_napari_viewer, tmp_plugin: DynamicPlugin, Widget
+):
+    """Test all supported widget types correctly instantiated and call processor."""
+
+    @tmp_plugin.contribute.widget(display_name='Widget')
+    def widget():
+        return Widget()
```

Comment:
I'm not sure this affects what you're trying to test, but I changed this to `autogenerate` because the different test widgets have different signatures and I think this should handle it with injection.

Reply:
@aganders3 can you please expand on this a bit? I don't think I understand (just for my own reference).

Reply:
Sorry, yeah I can try. What I meant here is that this test is parametrized on `Widget` for four types - `QWidget_example`, `Container_example`, `magic_widget_example`, and `callable_example`. Each of these is callable to create a widget, but the signature is different. For `QWidget_example` and `Container_example` they expect a viewer. `callable_example` and `magic_widget_example` each take no arguments.

This was kind of working before because (see my suggested changes) `__init__` for `QWidget_example` and `Container_example` were missing `self`, so they could be constructed without args.

Instead of putting logic in the test to determine how to construct the widget, I think (*I think*) `autogenerate=True` will just figure out how to do it, using dependency injection as needed.

Where I said:
> I'm not sure this affects what you're trying to test
that's because I think now it uses `magicgui` for all the examples, which I'm not sure is desirable. At least it might mean the `magic_widget_example` is redundant?

Reply:
Ahhh ok got it, thanks for the detailed response @aganders3. I think in this case we don't want to use `autogenerate`, because we want to make sure we support whatever widget constructors people might have in their plugins. However, I do think the callable we're decorating should return the `Widget` without calling it, because I don't think it should be returning an already constructed widget - @lucyleeow isn't it the command callable's job to build the widget?

Reply:
Yes, from the [docs](The widget specification requires that the plugin provide napari with a callable object that, when called, returns an instance of a widget.):

> The widget specification requires that the plugin provide napari with a callable object that, when called, returns an instance of a widget.

Reply:
I think the test (with the new `super`) will still work? It's meant to be able to handle the different signatures. I will add note to make this explicit.

Reply:
I'll need to check tomorrow but I think it's no longer returning an _instance_ of a widget. I don't know what happens in that case.

Reply:
I don't think this is testing properly now. To see what I mean try printing `print([(k, v) for k, v in viewer.window._dock_widgets.items()])` after executing the command. Without `autogenerate` this is empty, but with it you see the dock widget added (you need to set the side-effect for this as well). I think this also explains why nothing is leaked - no widget is created!

This works and I believe it tests what is expected - shorter and more readable to me:
```
@pytest.mark.parametrize(
    "Widget",
    [
        QWidget_example,
        Container_example,
        magic_widget_example,
        callable_example,
    ],
)
def test_widget_types_supported(
    make_napari_viewer,
    tmp_plugin: DynamicPlugin,
    Widget,
):
    """Test all supported widget types correctly instantiated and call processor."""
    # The 4 `Widget`s should represent varing widget constructors and signatures that
    # we want to support.
    tmp_plugin.contribute.widget(display_name='Widget')(Widget)

    app = get_app()
    viewer = make_napari_viewer()

    app.commands.execute_command('tmp_plugin:Widget')
    assert 'Widget (Temp Plugin)' in viewer.window._dock_widgets
```

Reply:
I still don't think we want to use `autoogenerate`, because we want to test it works without it? I will investigate.

But yes you are right that testing that the `add_dock_widget` processor is used *probably* does end up being the same as checking it was added to the window widget dict BUT it is less of a unit test in the latter case. I was aiming for a 'real' unit test that would specifically only test the `_get_widgets_submenu_actions` and processor part (parts we added in this PR).

I think there are some existing tests that check the function `add_dock_widget` works properly.

Reply:
The version I posted above doesn't use `autogenerate` now so I'm on board with that. I see what you mean about the...unit-ness. The nice thing about actually calling the fn is that it prevents the widget leak for me. I can look into how to clean up the widget if there's still a leak here.

Reply:
Ahhhh the widget is not added to `window._dock_widgets` because I mocked the function `add_dock_widget`. I think I will keep as is because I want a unit test, even though I agree it's probably the same in most cases.

Reply:
Okay yes my widget definition was not right! `tmp_plugin.contribute.widget(display_name='Widget')(Widget)` is also cleaner.

The widget leak does not happen for me, even with the `MagicMock`. I do not know how I got the leaks before now...

Reply:
Okay my testing got mixed up. I have added `side_effect` to the mock and that seems to fix the leak.